### ISSUE TITLE
Issue when checking for nil custom error

### ISSUE THREAD

**Issue by eduardboamba** [gotestyourself/gotest.tools#167]
Create a custom error by implementing the `error` interface (e.g. Error() function)
```
type MyCustomError struct {
    msg string
}
func (ce *MyCustomError) Error() string {
    return ce.string
}
```

In a test, checking for a nil MyCustomError reference using the `NilError` function will lead to unexpected behavior, such as:
```
var myErr *MyCustomError 
assert.NilError(t, myErr)
```
in this case, you receive a memory address error, because the above custom `Error()` function will get invoked, even though `ce` is `nil`.

If you "fix" your `Error()` function, such as:
```
func (ce *MyCustomError) Error() string {
    if ce != nil {
        return ce.string
    }
    return ""
}
```
then you get rid of the null pointer memory error, but the assertion will fail.

This happens because, in the `assert.go` file, at the type switch `switch check := comparison.(type)`, it resolves to `case error` branch.

Workaround: use `Assert(t, myErr == nil)` instead.

**Comment by dnephin** [comment]
I believe this is because of how interfaces work. Because you are passing a struct pointer, not an interface. https://www.pixelstech.net/article/1554553347-Be-careful-about-nil-check-on-interface-in-GoLang

I'm not sure if there is any way to fix this problem. `NilError` expects an interface value, not a struct pointer.

**Comment by eduardboamba** [comment]
I see, thank you for the reference. I understand now that the assert package is dealing with the cases where the `error` variable is never set, which results in a `nil` type. I am fine using the workaround I mentioned.
I was trying to come up with an acceptable solution/improvement, to cover checking implementations of the `error` interface as well; the only "working" way I found was to use the reflect package and change the type switch to:
```
	case error:
		if reflect.ValueOf(comparison).IsNil() {
			return true
		}
		msg := "error is not nil: "
		t.Log(format.WithCustomMessage(failureMessage+msg+check.Error(), msgAndArgs...))
```

**Comment by dnephin** [comment]
Thanks for reporting this issue! I've had a chance to look into it more now.

It turns out we had a test case for this (https://github.com/gotestyourself/gotest.tools/blob/172219a/assert/assert_test.go#L134-L135) but it was missing an assertion for the output. Oops! When the assert is added it does indeed fail.

I was able to fix it using the suggestion you made above, so thank you also for the fix!

I've pushed the fix in #171, and linked to a more official document explaining the issue: 
https://golang.org/doc/faq#nil_error

**Comment by dnephin** [comment]
Just a heads up that there was a problem with this fix. I've pushed a different fix in #187 
This will cause `NilError` to fail if passed a typed `error`, but it will no longer panic, it should report the type to indicate why it is not nil.
